src/transform/pre_reduce.rs

```diff
@@ -19,39 +22,131 @@ impl run::AsDef for InertDef {
   }
 }
 
+pub struct PreReduceStats {
+  pub rewrites: Rewrites,
+  pub errors: Vec<String>,
+}
+
 impl Book {
   /// Reduces the definitions in the book individually, except for the skipped
   /// ones.
   ///
   /// Defs that are not in the book are treated as inert defs.
-  pub fn pre_reduce(&mut self, skip: &dyn Fn(&str) -> bool, max_memory: usize, max_rwts: u64) -> Result<(), String> {
+  pub fn pre_reduce(&mut self, skip: &dyn Fn(&str) -> bool, max_memory: usize, max_rwts: u64) -> PreReduceStats {
+    /// State of the pre-reduction algorithm.
+    ///
+    /// Here's how it works:
+    /// - Each definition is visited in reverse topological order (dependencies
+    ///   then dependents)
+    /// - The definition is reduced in a [`run::Net`]
+    /// - The reduced [`run::Net`] is readback into an [`ast::Net`]
+    /// - The [`ast::Net`] is placed into the [`Book`], overriding the previous
+    ///   one.
+    /// - The [`ast::Net`] is encoded into a [`Vec<Instruction>`]
+    /// - The [`InterpretedDef`] corresponding to the definition is mutated
+    ///   in-place and its instructions are replaced with the generated
+    ///   [`Vec<Instruction>`]
+    ///
+    /// This is repeated with each definition in the book.
+    struct State<'a> {
+      book: &'a mut Book,
+
+      host: Host,
+      max_rwts: u64,
+
+      area: &'a Heap,
+      captured_redexes: Arc<Mutex<Vec<(run::Port, run::Port)>>>,
+
+      skip: &'a dyn Fn(&str) -> bool,
+      cycled: HashSet<String>,
+      reduced: HashSet<String>,
+
+      total_rewrites: Rewrites<u64>,
+    }
+    impl<'a> State<'a> {
+      fn visit_tree(&self, tree: &Tree, out: &mut Vec<String>) {
```

Comment:
IMO these `visit_{tree,net}` functions should just call `pre_reduce` rather than inserting into a vec.

Reply:
This is not possible right now as we're borrowing both `nam` and `net` in `pre_reduce` from the same `net` in `visit_{tree, net}`. (and `net` is `&mut` while `nam` is `&`). The approach we use will have to make the reference projection explicit. (maybe passing `book` as an argument and modifying it afterwards outside of `State`?)

Reply:
Ah yeah, that's fair. Pushed a commit with some changes that resolves this.